File: src/zoho_books_cli/commands/bank_transactions.py
"""`zb bank-transactions ...` — full coverage of /banktransactions.

Thin wrappers: CRUD, action verbs (match/unmatch/exclude/restore/uncategorize),
an 8-variant categorize sub-app, and the /banktransactions/bulk/statement
import subtree.

Several action verbs sit under /banktransactions/uncategorized/{id}/... per
Zoho's API layout; `unmatch` and `uncategorize` sit directly under
/banktransactions/{id}/... despite the naming asymmetry.
"""

from __future__ import annotations

import typer

from zoho_books_cli import config
from zoho_books_cli.client import ZohoBooksClient
from zoho_books_cli.commands import _shared

app = typer.Typer(
    help="Bank transaction operations (CRUD + match/categorize + statements).",
    no_args_is_help=True,
)
categorize_app = typer.Typer(
    help="Categorize an uncategorized bank transaction (one subcommand per target type).",
    no_args_is_help=True,
)
statements_app = typer.Typer(
    help="Bulk bank statement import operations.", no_args_is_help=True
)
app.add_typer(categorize_app, name="categorize")
app.add_typer(statements_app, name="statements")

BASE = "/banktransactions"
UNCAT = f"{BASE}/uncategorized"


@app.command("list")
def list_transactions(
    query: list[str] = typer.Option(
        None, "--query", "-q", help="Query params as key=value. May be repeated."
    ),
    page: int = typer.Option(None, "--page", help="Page number (1-indexed)."),
    per_page: int = typer.Option(None, "--per-page", help="Rows per page."),
):
    """List bank transactions. Returns one page plus page_context."""
    q = _shared.parse_query_pairs(query)
    if page is not None:
        q["page"] = str(page)
    if per_page is not None:
        q["per_page"] = str(per_page)
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.get(BASE, query=q)
    _shared.emit_list(resp, "banktransactions")


@app.command("create")
def create_transaction(
    body: str = typer.Option(..., "--body", "-b", help="JSON body. IDs must be strings."),
):
    """Create a bank transaction."""
    json_body = _shared.parse_body(body)
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.post(BASE, json_body=json_body)
    _shared.emit_object(resp)


@app.command("get")
def get_transaction(
    transaction_id: str = typer.Argument(..., help="Zoho Books bank_transaction_id."),
):
    """Get a single bank transaction by ID."""
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.get(f"{BASE}/{transaction_id}")
    _shared.emit_object(resp)


@app.command("update")
def update_transaction(
    transaction_id: str = typer.Argument(..., help="Zoho Books bank_transaction_id."),
    body: str = typer.Option(..., "--body", "-b", help="JSON body. IDs must be strings."),
):
    """Update a bank transaction by ID."""
    json_body = _shared.parse_body(body)
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.put(f"{BASE}/{transaction_id}", json_body=json_body)
    _shared.emit_object(resp)


@app.command("delete")
def delete_transaction(
    transaction_id: str = typer.Argument(..., help="Zoho Books bank_transaction_id."),
):
    """Delete a bank transaction by ID."""
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.delete(f"{BASE}/{transaction_id}")
    _shared.emit_action("transaction_id", transaction_id, resp)


@app.command("match")
def match_transaction(
    transaction_id: str = typer.Argument(..., help="Zoho Books bank_transaction_id."),
    body: str = typer.Option(
        ..., "--body", "-b", help="JSON body with transactions_to_be_matched. IDs must be strings."
    ),
):
    """Match an uncategorized transaction (POST /banktransactions/uncategorized/{id}/match)."""
    json_body = _shared.parse_body(body)
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.post(f"{UNCAT}/{transaction_id}/match", json_body=json_body)
    _shared.emit_action("transaction_id", transaction_id, resp)


@app.command("matches")
def list_matches(
    transaction_id: str = typer.Argument(..., help="Zoho Books bank_transaction_id."),
    query: list[str] = typer.Option(None, "--query", "-q", help="Query params as key=value."),
    page: int = typer.Option(None, "--page", help="Page number (1-indexed)."),
    per_page: int = typer.Option(None, "--per-page", help="Rows per page."),
):
    """List candidate matching transactions (GET /banktransactions/uncategorized/{id}/match)."""
    q = _shared.parse_query_pairs(query)
    if page is not None:
        q["page"] = str(page)
    if per_page is not None:
        q["per_page"] = str(per_page)
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.get(f"{UNCAT}/{transaction_id}/match", query=q)
    _shared.emit_list(resp, "matching_transactions")


@app.command("unmatch")
def unmatch_transaction(
    transaction_id: str = typer.Argument(..., help="Zoho Books bank_transaction_id."),
):
    """Unmatch a matched transaction (POST /banktransactions/{id}/unmatch)."""
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.post(f"{BASE}/{transaction_id}/unmatch")
    _shared.emit_action("transaction_id", transaction_id, resp)


@app.command("exclude")
def exclude_transaction(
    transaction_id: str = typer.Argument(..., help="Zoho Books bank_transaction_id."),
):
    """Exclude a transaction (POST /banktransactions/uncategorized/{id}/exclude)."""
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.post(f"{UNCAT}/{transaction_id}/exclude")
    _shared.emit_action("transaction_id", transaction_id, resp)


@app.command("restore")
def restore_transaction(
    transaction_id: str = typer.Argument(..., help="Zoho Books bank_transaction_id."),
):
    """Restore an excluded transaction (POST /banktransactions/uncategorized/{id}/restore)."""
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.post(f"{UNCAT}/{transaction_id}/restore")
    _shared.emit_action("transaction_id", transaction_id, resp)


@app.command("uncategorize")
def uncategorize_transaction(
    transaction_id: str = typer.Argument(..., help="Zoho Books bank_transaction_id."),
):
    """Uncategorize a categorized transaction (POST /banktransactions/{id}/uncategorize)."""
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.post(f"{BASE}/{transaction_id}/uncategorize")
    _shared.emit_action("transaction_id", transaction_id, resp)


# --- categorize sub-app ------------------------------------------------------


def _categorize(transaction_id: str, subpath: str, body: str | None) -> None:
    json_body = _shared.parse_body(body)
    path = f"{UNCAT}/{transaction_id}/categorize"
    if subpath:
        path = f"{path}/{subpath}"
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.post(path, json_body=json_body)
    _shared.emit_action("transaction_id", transaction_id, resp)


@categorize_app.command("generic")
def categorize_generic(
    transaction_id: str = typer.Argument(..., help="Zoho Books bank_transaction_id."),
    body: str = typer.Option(..., "--body", "-b", help="JSON body. IDs must be strings."),
):
    """Manual categorization (POST /banktransactions/uncategorized/{id}/categorize)."""
    _categorize(transaction_id, "", body)


@categorize_app.command("expense")
def categorize_expense(
    transaction_id: str = typer.Argument(..., help="Zoho Books bank_transaction_id."),
    body: str = typer.Option(..., "--body", "-b", help="JSON body. IDs must be strings."),
):
    """Categorize as expense (POST .../categorize/expenses)."""
    _categorize(transaction_id, "expenses", body)


@categorize_app.command("vendor-payment")
def categorize_vendor_payment(
    transaction_id: str = typer.Argument(..., help="Zoho Books bank_transaction_id."),
    body: str = typer.Option(..., "--body", "-b", help="JSON body. IDs must be strings."),
):
    """Categorize as vendor payment (POST .../categorize/vendorpayments)."""
    _categorize(transaction_id, "vendorpayments", body)


@categorize_app.command("customer-payment")
def categorize_customer_payment(
    transaction_id: str = typer.Argument(..., help="Zoho Books bank_transaction_id."),
    body: str = typer.Option(..., "--body", "-b", help="JSON body. IDs must be strings."),
):
    """Categorize as customer payment (POST .../categorize/customerpayments)."""
    _categorize(transaction_id, "customerpayments", body)


@categorize_app.command("credit-note-refund")
def categorize_credit_note_refund(
    transaction_id: str = typer.Argument(..., help="Zoho Books bank_transaction_id."),
    body: str = typer.Option(..., "--body", "-b", help="JSON body. IDs must be strings."),
):
    """Categorize as credit note refund (POST .../categorize/creditnoterefunds)."""
    _categorize(transaction_id, "creditnoterefunds", body)


@categorize_app.command("vendor-credit-refund")
def categorize_vendor_credit_refund(
    transaction_id: str = typer.Argument(..., help="Zoho Books bank_transaction_id."),
    body: str = typer.Option(..., "--body", "-b", help="JSON body. IDs must be strings."),
):
    """Categorize as vendor credit refund (POST .../categorize/vendorcreditrefunds)."""
    _categorize(transaction_id, "vendorcreditrefunds", body)


@categorize_app.command("payment-refund")
def categorize_payment_refund(
    transaction_id: str = typer.Argument(..., help="Zoho Books bank_transaction_id."),
    body: str = typer.Option(..., "--body", "-b", help="JSON body. IDs must be strings."),
):
    """Categorize as customer payment refund (POST .../categorize/paymentrefunds)."""
    _categorize(transaction_id, "paymentrefunds", body)


@categorize_app.command("vendor-payment-refund")
def categorize_vendor_payment_refund(
    transaction_id: str = typer.Argument(..., help="Zoho Books bank_transaction_id."),
    body: str = typer.Option(..., "--body", "-b", help="JSON body. IDs must be strings."),
):
    """Categorize as vendor payment refund (POST .../categorize/vendorpaymentrefunds)."""
    _categorize(transaction_id, "vendorpaymentrefunds", body)


# --- statements sub-app ------------------------------------------------------
# Statements are NOT under /banktransactions in Zoho's API — import is a
# top-level /bankstatements resource, and the get/delete operations live under
# /bankaccounts/{account_id}/statement/... We expose them under
# `bank-transactions statements` for user ergonomics since that's the natural
# conceptual grouping.


@statements_app.command("import")
def statements_import(
    body: str = typer.Option(
        ...,
        "--body",
        "-b",
        help="JSON body. Must include account_id, from_date, to_date, and file_content. "
        "IDs must be strings.",
    ),
):
    """Import a bank/credit-card statement (POST /bankstatements)."""
    json_body = _shared.parse_body(body)
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.post("/bankstatements", json_body=json_body)
    _shared.emit_object(resp)


@statements_app.command("last-imported")
def statements_last_imported(
    account_id: str = typer.Argument(..., help="Zoho Books bank account_id."),
):
    """Fetch the last imported statement for an account.

    (GET /bankaccounts/{account_id}/statement/lastimported)
    """
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.get(f"/bankaccounts/{account_id}/statement/lastimported")
    _shared.emit_object(resp)


@statements_app.command("delete")
def statements_delete(
    account_id: str = typer.Argument(..., help="Zoho Books bank account_id."),
    statement_id: str = typer.Argument(..., help="Zoho Books statement_id to delete."),
):
    """Delete a specific imported statement.

    (DELETE /bankaccounts/{account_id}/statement/{statement_id})
    """
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.delete(f"/bankaccounts/{account_id}/statement/{statement_id}")
    _shared.emit_action("statement_id", statement_id, resp)

File: src/commands/bank_transactions.rs
//! `zb bank-transactions` — CRUD + match/unmatch/exclude/restore/uncategorize
//! + categorize sub-app + statements sub-app.
//!
//! Generic categorize body (verified against Zoho v3 docs): the manual
//! `categorize generic` endpoint is double-entry and requires
//! `transaction_type` + `from_account_id` + `to_account_id` + `amount` +
//! `date`. `transaction_type` ∈ {deposit, refund (credit-card only),
//! transfer_fund, card_payment, sales_without_invoices, expense_refund,
//! owner_contribution, interest_income, other_income, owner_drawings,
//! sales_return}. expense / vendor_payment / customer_payment are rejected
//! here and route through the dedicated categorize sub-commands instead.

use clap::{Args, Subcommand};

use crate::cli::Ctx;
use crate::commands::common::{self, BodyArgs, ListArgs};
use crate::errors::Result;

const BASE: &str = "/banktransactions";
const UNCAT_PREFIX: &str = "/banktransactions/uncategorized";
const TID: &str = "transaction_id";

#[derive(Args, Debug)]
pub struct Cmd {
    #[command(subcommand)]
    pub sub: Sub,
}

#[derive(Subcommand, Debug)]
pub enum Sub {
    List(ListArgs),
    Create(BodyArgs),
    Get(IdArgs),
    Update(UpdateArgs),
    Delete(IdArgs),
    /// Match an uncategorized transaction (POST .../uncategorized/{id}/match).
    Match(MatchArgs),
    /// List candidate matching transactions (GET .../uncategorized/{id}/match).
    Matches(MatchesArgs),
    /// Unmatch a matched transaction (POST /banktransactions/{id}/unmatch).
    Unmatch(IdArgs),
    /// Exclude a transaction (POST .../uncategorized/{id}/exclude).
    Exclude(IdArgs),
    /// Restore an excluded transaction (POST .../uncategorized/{id}/restore).
    Restore(IdArgs),
    /// Uncategorize a categorized transaction.
    Uncategorize(IdArgs),
    /// Categorize sub-app (one subcommand per target type).
    Categorize(CategorizeCmd),
    /// Bulk bank statement import operations.
    Statements(StatementsCmd),
}

#[derive(Args, Debug)]
pub struct IdArgs {
    pub transaction_id: String,
}

#[derive(Args, Debug)]
pub struct UpdateArgs {
    pub transaction_id: String,
    #[command(flatten)]
    pub body: BodyArgs,
}

#[derive(Args, Debug)]
pub struct MatchArgs {
    pub transaction_id: String,
    #[command(flatten)]
    pub body: BodyArgs,
}

#[derive(Args, Debug)]
pub struct MatchesArgs {
    pub transaction_id: String,
    #[command(flatten)]
    pub list: ListArgs,
}

#[derive(Args, Debug)]
pub struct CategorizeCmd {
    #[command(subcommand)]
    pub sub: CategorizeSub,
}

#[derive(Subcommand, Debug)]
pub enum CategorizeSub {
    /// Manual ("generic") categorization (POST .../categorize).
    ///
    /// This posts a double-entry transaction, so the body needs BOTH accounts
    /// plus the type, amount, and date. A single `account_id` is NOT a valid
    /// field (that mistake yields "Invalid value passed for Transaction Type"
    /// or "Account does not exist"):
    ///
    ///   --body '{"transaction_type":"other_income",
    ///            "from_account_id":"<source_account_id>",
    ///            "to_account_id":"<bank_account_id>",
    ///            "amount":100.0,"date":"YYYY-MM-DD"}'
    ///
    /// `from_account_id` is where the money came from; `to_account_id` is where
    /// it landed. For money arriving in the bank (deposit, interest_income,
    /// other_income, owner_contribution, sales_without_invoices) the bank
    /// account is `to_account_id`; for money leaving it (owner_drawings,
    /// card_payment, transfer_fund) the bank account is `from_account_id`.
    ///
    /// transaction_type is one of: deposit, refund (credit-card accounts only),
    /// transfer_fund, card_payment, sales_without_invoices, expense_refund,
    /// owner_contribution, interest_income, other_income, owner_drawings,
    /// sales_return. Categorizing as an expense, vendor payment, or customer
    /// payment is NOT supported here: use `categorize expense`,
    /// `categorize vendor-payment`, or `categorize customer-payment`.
    Generic(CategorizeArgs),
    /// POST .../categorize/expenses
    Expense(CategorizeArgs),
    /// POST .../categorize/vendorpayments
    #[command(name = "vendor-payment")]
    VendorPayment(CategorizeArgs),
    /// POST .../categorize/customerpayments
    #[command(name = "customer-payment")]
    CustomerPayment(CategorizeArgs),
    /// POST .../categorize/creditnoterefunds
    #[command(name = "credit-note-refund")]
    CreditNoteRefund(CategorizeArgs),
    /// POST .../categorize/vendorcreditrefunds
    #[command(name = "vendor-credit-refund")]
    VendorCreditRefund(CategorizeArgs),
    /// POST .../categorize/paymentrefunds
    #[command(name = "payment-refund")]
    PaymentRefund(CategorizeArgs),
    /// POST .../categorize/vendorpaymentrefunds
    #[command(name = "vendor-payment-refund")]
    VendorPaymentRefund(CategorizeArgs),
}

#[derive(Args, Debug)]
pub struct CategorizeArgs {
    pub transaction_id: String,
    #[command(flatten)]
    pub body: BodyArgs,
}

#[derive(Args, Debug)]
pub struct StatementsCmd {
    #[command(subcommand)]
    pub sub: StatementsSub,
}

#[derive(Subcommand, Debug)]
pub enum StatementsSub {
    /// Import a bank/credit-card statement (POST /bankstatements).
    Import(BodyArgs),
    /// Fetch the last imported statement for an account.
    #[command(name = "last-imported")]
    LastImported(AccountIdArgs),
    /// Delete a specific imported statement.
    Delete(StatementDeleteArgs),
}

#[derive(Args, Debug)]
pub struct AccountIdArgs {
    pub account_id: String,
}

#[derive(Args, Debug)]
pub struct StatementDeleteArgs {
    pub account_id: String,
    pub statement_id: String,
}

pub fn run(cmd: Cmd, ctx: &mut Ctx) -> Result<()> {
    match cmd.sub {
        Sub::List(args) => common::list(ctx, BASE, &args, "banktransactions"),
        Sub::Create(args) => common::create(ctx, BASE, &args),
        Sub::Get(args) => common::get(ctx, &format!("{BASE}/{}", args.transaction_id)),
        Sub::Update(args) => {
            common::update(ctx, &format!("{BASE}/{}", args.transaction_id), &args.body)
        }
        Sub::Delete(args) => {
            let path = format!("{BASE}/{}", args.transaction_id);
            common::delete(ctx, &path, TID, &args.transaction_id)
        }
        Sub::Match(args) => {
            let path = format!("{UNCAT_PREFIX}/{}/match", args.transaction_id);
            common::action_with_body(ctx, &path, &args.body, TID, &args.transaction_id)
        }
        Sub::Matches(args) => {
            let path = format!("{UNCAT_PREFIX}/{}/match", args.transaction_id);
            common::list(ctx, &path, &args.list, "matching_transactions")
        }
        Sub::Unmatch(args) => {
            let path = format!("{BASE}/{}/unmatch", args.transaction_id);
            common::action(ctx, &path, TID, &args.transaction_id)
        }
        Sub::Exclude(args) => {
            let path = format!("{UNCAT_PREFIX}/{}/exclude", args.transaction_id);
            common::action(ctx, &path, TID, &args.transaction_id)
        }
        Sub::Restore(args) => {
            let path = format!("{UNCAT_PREFIX}/{}/restore", args.transaction_id);
            common::action(ctx, &path, TID, &args.transaction_id)
        }
        Sub::Uncategorize(args) => {
            let path = format!("{BASE}/{}/uncategorize", args.transaction_id);
            common::action(ctx, &path, TID, &args.transaction_id)
        }
        Sub::Categorize(c) => match c.sub {
            CategorizeSub::Generic(args) => categorize(ctx, &args, ""),
            CategorizeSub::Expense(args) => categorize(ctx, &args, "expenses"),
            CategorizeSub::VendorPayment(args) => categorize(ctx, &args, "vendorpayments"),
            CategorizeSub::CustomerPayment(args) => categorize(ctx, &args, "customerpayments"),
            CategorizeSub::CreditNoteRefund(args) => categorize(ctx, &args, "creditnoterefunds"),
            CategorizeSub::VendorCreditRefund(args) => {
                categorize(ctx, &args, "vendorcreditrefunds")
            }
            CategorizeSub::PaymentRefund(args) => categorize(ctx, &args, "paymentrefunds"),
            CategorizeSub::VendorPaymentRefund(args) => {
                categorize(ctx, &args, "vendorpaymentrefunds")
            }
        },
        Sub::Statements(s) => match s.sub {
            StatementsSub::Import(args) => common::create(ctx, "/bankstatements", &args),
            StatementsSub::LastImported(args) => common::get(
                ctx,
                &format!("/bankaccounts/{}/statement/lastimported", args.account_id),
            ),
            StatementsSub::Delete(args) => {
                let path = format!(
                    "/bankaccounts/{}/statement/{}",
                    args.account_id, args.statement_id
                );
                common::delete(ctx, &path, "statement_id", &args.statement_id)
            }
        },
    }
}

fn categorize(ctx: &mut Ctx, args: &CategorizeArgs, suffix: &str) -> Result<()> {
    let path = if suffix.is_empty() {
        format!("{UNCAT_PREFIX}/{}/categorize", args.transaction_id)
    } else {
        format!("{UNCAT_PREFIX}/{}/categorize/{suffix}", args.transaction_id)
    };
    common::action_with_body(ctx, &path, &args.body, TID, &args.transaction_id)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::cli::Ctx;

    #[test]
    fn list_targets_banktransactions() {
        let mut server = mockito::Server::new();
        let m = server
            .mock("GET", "/books/v3/banktransactions")
            .match_query(mockito::Matcher::Any)
            .with_status(200)
            .with_body(r#"{"banktransactions":[]}"#)
            .create();
        let mut ctx = Ctx::new_for_test(&server.url());
        run(
            Cmd {
                sub: Sub::List(ListArgs::default()),
            },
            &mut ctx,
        )
        .unwrap();
        m.assert();
    }
}
